pkg/controllers/machinesetsync/machineset_sync_controller_test.go

```diff
@@ -156,56 +155,280 @@ var _ = Describe("With a running MachineSetSync controller", func() {
+					Expect(k8sClient.Create(ctx, capiMachineSet)).Should(Succeed())
+					Expect(k8sClient.Create(ctx, mapiMachineSet)).Should(Succeed())
+
+					By("Setting the AuthoritativeAPI to ClusterAPI")
+					Eventually(k.UpdateStatus(mapiMachineSet, func() {
+						mapiMachineSet.Status.AuthoritativeAPI = machinev1beta1.MachineAuthorityClusterAPI
+					})).Should(Succeed())
+				})
+
+				It("should update the synchronized condition on the MAPI MachineSet", func() {
+					Eventually(k.Object(mapiMachineSet), timeout).Should(
+						SatisfyAll(
+							HaveField("Status.Conditions", ContainElement(
+								SatisfyAll(
+									HaveField("Type", Equal(consts.SynchronizedCondition)),
+									HaveField("Status", Equal(corev1.ConditionFalse)),
+									HaveField("Severity", Equal(machinev1beta1.ConditionSeverityError)),
+									HaveField("Reason", Equal("FailedToConvertCAPIMachineSetToMAPI")),
+								))),
+						))
+
+				})
+			})
+
 		})
 
-		It("should update the replica count", func() {
-			Eventually(k.Object(mapiMachineSet), timeout).Should(
-				HaveField("Spec.Replicas", Equal(ptr.To(int32(4)))),
-			)
+		Context("when the conversion has warnings", func() {})
+
+		Context("when the CAPI MachineSet does not exist", func() {
+			BeforeEach(func() {
+				By("Creating the CAPI and MAPI MachineSets")
+				mapiMachineSet = mapiMachineSetBuilder.Build()
+				Expect(k8sClient.Create(ctx, mapiMachineSet)).Should(Succeed())
+
+				By("Setting the AuthoritativeAPI to ClusterAPI")
+				Eventually(k.UpdateStatus(mapiMachineSet, func() {
+					mapiMachineSet.Status.AuthoritativeAPI = machinev1beta1.MachineAuthorityClusterAPI
+				})).Should(Succeed())
+			})
+
+			It("should update the synchronized condition on the MAPI MachineSet", func() {
+				Eventually(k.Object(mapiMachineSet), timeout).Should(
+					SatisfyAll(
+						HaveField("Status.Conditions", ContainElement(
+							SatisfyAll(
+								HaveField("Type", Equal(consts.SynchronizedCondition)),
+								HaveField("Status", Equal(corev1.ConditionFalse)),
+								HaveField("Severity", Equal(machinev1beta1.ConditionSeverityError)),
+								HaveField("Reason", Equal("FailedToGetCAPIMachineSet")),
+								HaveField("Message", Equal("error synchronizing CAPI to MAPI, CAPI machineset not found")),
+							))),
+					))
+
+			})
+
 		})
+
+		Context("When the MAPI MachineAuthority is set to Migrating", func() {
+			BeforeEach(func() {
+				By("Creating the CAPI and MAPI MachineSets")
+				mapiMachineSet = mapiMachineSetBuilder.Build()
+				capiMachineSet = capiMachineSetBuilder.Build()
+
+				Expect(k8sClient.Create(ctx, capiMachineSet)).Should(Succeed())
+				Expect(k8sClient.Create(ctx, mapiMachineSet)).Should(Succeed())
+
+				By("Setting the AuthoritativeAPI to Migrating")
+				Eventually(k.UpdateStatus(mapiMachineSet, func() {
+					mapiMachineSet.Status.AuthoritativeAPI = machinev1beta1.MachineAuthorityMigrating
+				})).Should(Succeed())
+			})
+
+			It("should not make any changes", func() {
```

Comment:
What's this test actually doing? Are we creating a scenario where it would normally be trying to update one side or the other, and now it's not? How is it obvious that it should only have been trying to update the mapi side?

Reply:
Ah I'd forgotten to add differences between the two machinesets. 🤦 

I've updated this so we have n replicas on one, and n+3 on the other. In migrating we should see both remain the same, but if we reconcile in either direction we'll get a different resource version and fail the test.